README.md
[![Build Status](https://travis-ci.org/issm/p5-String-Incremental.png?branch=master)](https://travis-ci.org/issm/p5-String-Incremental)
# NAME

String::Incremental - incremental string with your rule

# SYNOPSIS

    use String::Incremental;

    my $str = String::Incremental->new(
        format => 'foo-%2=-%=',
        orders => [
            [0..2],
            'abcd',
        ],
    );

    # or

    use String::Incremental qw( incremental_string );

    my $str = incremental_string(
        'foo-%2=-%=',
        [0..2],
        'abcd',
    );

    print "$str";  # -> 'foo-00-a'

    $str++; $str++; $str++;
    print "$str";  # -> 'foo-00-d'

    $str++;
    print "$str";  # -> 'foo-01-a'

    $str->set( 'foo-22-d' );
    print "$str";  # -> 'foo-22-d';
    $str++;  # dies, cannot ++ any more

# DESCRIPTION

String::Incremental is ...

# CONSTRUCTORS

- new( %args ) : String::Incremental

    format: Str

    orders: ArrayRef

# METHODS

- as\_string() : Str

    returns "current" string.

    following two variables are equivalent:

        my $a = $str->as_string();
        my $b = "$str";

- set( $val : Str ) : String::Incremental

    sets to $val.

    tying with String::Incremental, assignment syntax is available as synonym of this method:

        tie my $str, 'String::Incremental', (
            format => 'foo-%2=-%=',
            orders => [ [0..2], 'abcd' ],
        );

        $str = 'foo-22-d';  # same as `$str->set( 'foo-22-d' )`
        print "$str";  # -> 'foo-22-d';

- increment() : Str

    increases positional state of order and returns its character.

    following two operation are equivalent:

        $str->increment();
        $str++;

- decrement() : Str

    decreases positional state of order and returns its character.

    following two operation are equivalent:

        $str->decrement();
        $str--;

# FUNCTIONS

- incremental\_string( $format, @orders ) : String::Incremental

    another way to construct String::Incremental instance.

    this function is not exported automatically, you need to export manually:

        use String::Incremental qw( incremental_string );

# LICENSE

Copyright (C) issm.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

issm <issmxx@gmail.com>


Changes
Revision history for Perl extension String-Incremental

{{$NEXT}}

0.01 2014-07-17T15:02:23Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'String-Incremental',
    module_name     => 'String::Incremental',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,


);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/String/Incremental/Char.pm
package String::Incremental::Char;
use 5.008005;
use warnings;
use Mouse;
use Data::Validator;
use MouseX::Types::Mouse qw( Bool Int );
use String::Incremental::Types qw( Char CharOrderStr CharOrderArrayRef is_CharOrderStr );
use Try::Tiny;

use overload (
    '""' => \&as_string,
    '++' => \&increment,
    '--' => \&decrement,
    '='  => sub { $_[0] },
);

has 'order'  => ( is  => 'ro', isa => CharOrderArrayRef );
has 'upper'  => ( is  => 'ro', isa => __PACKAGE__, predicate => 'has_upper' );
has '__i'    => ( is => 'rw', isa => Int );
has '__size' => ( is => 'ro', isa => Int );

sub BUILDARGS {
    my ($class, %args) = @_;
    my $v = Data::Validator->new(
        order => { isa => CharOrderStr|CharOrderArrayRef },
        upper => { isa => __PACKAGE__, optional => 1 },
        set   => { isa => Char, optional => 1 },
        __i   => { isa => Int, default => 0 },  # for internal use
    );
    %args = %{$v->validate( \%args )};

    if ( is_CharOrderStr( $args{order} ) ) {
        $args{order} = [ split //, $args{order} ];
    }

    %args = (
        %args,
        __size => 0 + @{$args{order}},
    );

    if ( $args{__i} >= $args{__size} ) {
        my $msg = '__i: should be less than size of "order"';
        die $msg;
    }

    return \%args;
}

sub BUILD {
    my ($self, $args) = @_;
    if ( exists $args->{set} ) {
        $self->set( $args->{set} );
    }
}

sub as_string {
    my ($self) = @_;
    return $self->order->[ $self->__i ];
}

sub set {
    my ($self, $ch, $opts) = @_;
    my $v_opts = Data::Validator->new(
        test => { isa => Bool, default => 0 },
    );
    $opts = $v_opts->validate( %{$opts || {}} );

    unless ( defined $ch ) {
        die 'value to set must be specified';
    }
    unless ( $ch =~ /^.$/ ) {
        die 'must be specified as a character';
    }
    my ($i, $found) = ( 0, 0 );
    for ( @{$self->order} ) {
        if ( $_ eq $ch ) {
            $found = 1;
            last;
        }
        $i++;
    }
    unless ( $found ) {
        my $msg = srintf( '"%s" is not in order', $ch );
        die $msg;
    }
    unless ( $opts->{test} ) {
        $self->__i( $i );
    }
    return "$self";
}

sub increment {
    my ($self) = @_;
    my ($i_try, $ch);
    $i_try = $self->__i + 1;

    if ( $i_try >= $self->__size ) {
        if ( $self->has_upper() ) {
            my $upper = $self->upper;
            try {
                $upper->increment();
                $i_try = 0;
            } catch {
                my ($msg) = @_;
                die $msg;
            };
        }
        else {
            my $msg = 'cannot increment';
            die $msg;
        }
    }

    $self->__i( $i_try );
    return $self->as_string();
}

sub decrement {
    my ($self) = @_;
    my ($i_try, $ch);
    $i_try = $self->__i - 1;

    if ( $i_try < 0 ) {
        if ( $self->has_upper() ) {
            my $upper = $self->upper;
            try {
                $upper->decrement();
                $i_try = $self->__size - 1;
            } catch {
                my ($msg) = @_;
                die $msg;
            };
        }
        else {
            my $msg = 'cannot decrement';
            die $msg;
        }
    }

    $self->__i( $i_try );
    return $self->as_string();
}

__PACKAGE__->meta->make_immutable();
__END__

=encoding utf-8

=head1 NAME

String::Incremental::Char

=head1 SYNOPSIS

    use String::Incremental::Char;

    my $ch = String::Incremental::Char->new( order => 'abcd' );

    print "$ch";  # -> 'a';

    $ch++; $ch++; $ch++;
    print "$ch";  # -> 'd';

    $ch++;  # dies

    my $ch1 = String::Incremental::Char->new( order => ['a'..'c'] );
    my $ch2 = String::Incremental::Char->new( order => ['x'..'z'], upper => $ch1 );

    print "${ch1}${ch2}";  # -> ax

    $ch2++; $ch2++;
    print "${ch1}${ch2}";  # -> az

    $ch2++;
    print "${ch1}${ch2}";  # -> bx

    $ch1++;
    print "${ch1}${ch2}";  # -> cx

    ...

    print "${ch1}${ch2}";  # -> cz
    $ch2++;  # dies


=head1 DESCRIPTION

String::Incremental::Char is ...

=head1 CONSTRUCTORS

=over 4

=item new( %args ) : String::Incremental::Char

%args:

order : Str|ArrayRef

  incrementation rule

upper : String::Incremental::Char

  upper-digit char as String::Incremental::Char instance

=back

=head1 METHODS

=over 4

=item as_string() : Str

returns "current" character.

following two variables are equivalent:

    my $a = $ch->as_string();
    my $b = "$ch";

=item set( $val, \%opts ) : String::Incremental::Char

sets "current" state as $val.

if $opts->{test} is true, "current" state is not update, only returns or dies.

=item increment() : Str

increases position of order and returns its character.

following two operation are equivalent:

    $ch->increment();
    $ch++;

=item decrement() : Str

decreases position of order and returns its character.

following two operation are equivalent:

    $ch->decrement();
    $ch--;

=back


=head1 LICENSE

Copyright (C) issm.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

issm E<lt>issmxx@gmail.comE<gt>

=cut


lib/String/Incremental/Types.pm
package String::Incremental::Types;
use 5.008005;
use strict;
use warnings;
use MouseX::Types -declare => [qw(
    CharOrderStr
    CharOrderArrayRef
)];
use MouseX::Types::Mouse qw( Str ArrayRef );

subtype CharOrderStr, as Str, where {
    my $val = shift;
    my %c;
    ( grep $c{$_}++, ( split //, $val ) ) ? 0 : 1;
};

subtype CharOrderArrayRef, as ArrayRef, where {
    my $val = shift;
    my %c;
    ( grep $c{$_}++, @$val ) ? 0 : 1;
};

1;
__END__

=encoding utf-8

=head1 NAME

String::Incremental::Types

=head1 SYNOPSIS

    use String::Incremental::Types qw( CharOrderStr CharOrderArrayref );

=head1 LICENSE

Copyright (C) issm.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

issm E<lt>issmxx@gmail.comE<gt>

=cut


lib/String/Incremental/String.pm
package String::Incremental::String;
use 5.008005;
use warnings;
use Mouse;
use Data::Validator;
use MouseX::Types::Mouse qw( Str CodeRef is_CodeRef );

use overload (
    '""' => \&as_string,
    '='  => sub { $_[0] },
);

has 'format' => ( is => 'ro', isa => Str );
has 'value' => ( is => 'ro', isa => CodeRef|Str );

sub BUILDARGS {
    my ($class, %args) = @_;
    my $v = Data::Validator->new(
        format => { isa => Str },
        value  => { isa => CodeRef|Str },
    );
    %args = %{$v->validate( \%args )};
    return \%args;
}

sub as_string {
    my ($self) = @_;
    my $val = $self->value;
    if ( is_CodeRef( $val ) ) { $val = $val->() }
    return sprintf( $self->format, $val );
}

__PACKAGE__->meta->make_immutable();
__END__

=encoding utf-8

=head1 NAME

String::Incremental::String

=head1 SYNOPSIS

    use String::Incremental::String;

    my $str = String::Incremental::String->new(
        format => '%04s',
        value  => sub { (localtime)[5] - 100 },
    );

    print "$str";  # -> '0014'


=head1 DESCRIPTION

String::Incremental::String is ...


=head1 CONSTRUCTORS

=over 4

=item new( %args ) : String::Incremental::String

%args:

format : Str

value : CodeRef|Str

=back


=head1 METHODS

=over 4

=item as_string() : Str

returns "current" string.

following two variables are equivalent:

    my $a = $str->as_string();
    my $b = "$str";

=back


=head1 LICENSE

Copyright (C) issm.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

issm E<lt>issmxx@gmail.comE<gt>

=cut


lib/String/Incremental.pm
package String::Incremental;
use 5.008005;
use strict;
use warnings;
use Mouse;
use MouseX::Types::Mouse qw( Str ArrayRef is_Str );
use String::Incremental::Types qw( Char );
use String::Incremental::FormatParser;
use String::Incremental::Char;
use Data::Validator;
use Try::Tiny;

use overload (
    '""' => \&as_string,
    '++' => \&increment,
    '--' => \&decrement,
    '='  => sub { $_[0] },
);

extends qw( Exporter Tie::Scalar );

our $VERSION = "0.01";

our @EXPORT_OK = qw( incremental_string );

has 'format' => ( is => 'ro', isa => Str );
has 'items'  => ( is => 'ro', isa => ArrayRef );
has 'chars'  => ( is => 'ro', isa => ArrayRef['String::Incremental::Char'] );

sub BUILDARGS {
    my ($class, %args) = @_;
    my $v = Data::Validator->new(
        format => { isa => Str },
        orders => { isa => ArrayRef, default => [] },
    );
    %args = %{$v->validate( \%args )};

    my $p = String::Incremental::FormatParser->new( $args{format}, @{$args{orders}} );

    return +{
        format => $p->format,
        items  => $p->items,
        chars  => [ grep $_->isa( __PACKAGE__ . '::Char' ), @{$p->items} ],
    };
}

sub incremental_string {
    my ($format, @orders) = @_;
    return __PACKAGE__->new( format => $format, orders => \@orders );
}

sub char {
    my ($self, $i) = @_;
    my $ch;
    unless ( defined $i ) {
        die 'index to set must be specified';
    }
    unless ( $i =~ /^\d+$/ ) {
        die 'must be specified as Int';
    }
    unless ( defined ( $ch = $self->chars->[$i] ) ) {
        die 'out of index';
    }
    return $ch;
}

sub as_string {
    my ($self) = @_;
    my @vals = map "$_", @{$self->items};
    return sprintf( $self->format, @vals );
}

sub set {
    my $v = Data::Validator->new(
        val => { isa => Str },
    )->with( 'Method', 'StrictSequenced' );
    my ($self, $args) = $v->validate( @_ );

    my @ch = $self->_extract_incremental_chars( $args->{val} );
    for ( my $i = 0; $i < @ch; $i++ ) {
        my $char = $self->char( $i );
        $char->set( $ch[$i] );
    }

    return "$self";
}

sub increment {
    my ($self) = @_;
    my ($last_ch) = grep $_->isa( __PACKAGE__ . '::Char' ), reverse @{$self->items};
    if ( defined $last_ch ) {
        $last_ch++;
    }
    return "$self";
}

sub decrement {
    my ($self) = @_;
    my ($last_ch) = grep $_->isa( __PACKAGE__ . '::Char' ), reverse @{$self->items};
    if ( defined $last_ch ) {
        $last_ch--;
    }
    return "$self";
}

sub re {
    my ($self) = @_;
    my ($re, @re);

    @re = map {
        my $i = $_;
        my $_re = $i->re();
        my $ref = ref $_;
        $ref eq __PACKAGE__ . '::Char' ? "(${_re})" : $_re;
    } @{$self->items};

    (my $fmt = $self->format) =~ s/%(?:\d+(?:\.?\d+)?)?\S/\%s/g;
    $re = sprintf $fmt, @re;

    return qr/^(${re})$/;
}

sub _extract_incremental_chars {
    my $v = Data::Validator->new(
        val => { isa => Str },
    )->with( 'Method', 'StrictSequenced' );
    my ($self, $args) = $v->validate( @_ );
    my @ch;

    (my $match, @ch) = $args->{val} =~ $self->re();
    unless ( defined $match ) {
        my $msg = 'specified value does not match with me';
        die $msg;
    }

    return wantarray ? @ch : \@ch;
}

sub TIESCALAR {
    my ($class, @args) = @_;
    return $class->new( @args );
}

sub FETCH { $_[0] }

sub STORE {
    my ($self, @args) = @_;
    if ( ref( $args[0] ) eq '' ) {  # ignore when ++/--
        $self->set( @args );
    }
}

__PACKAGE__->meta->make_immutable();
__END__

=encoding utf-8

=head1 NAME

String::Incremental - incremental string with your rule

=head1 SYNOPSIS

    use String::Incremental;

    my $str = String::Incremental->new(
        format => 'foo-%2=-%=',
        orders => [
            [0..2],
            'abcd',
        ],
    );

    # or

    use String::Incremental qw( incremental_string );

    my $str = incremental_string(
        'foo-%2=-%=',
        [0..2],
        'abcd',
    );

    print "$str";  # -> 'foo-00-a'

    $str++; $str++; $str++;
    print "$str";  # -> 'foo-00-d'

    $str++;
    print "$str";  # -> 'foo-01-a'

    $str->set( 'foo-22-d' );
    print "$str";  # -> 'foo-22-d';
    $str++;  # dies, cannot ++ any more

=head1 DESCRIPTION

String::Incremental is ...

=head1 CONSTRUCTORS

=over 4

=item new( %args ) : String::Incremental

format: Str

orders: ArrayRef

=back

=head1 METHODS

=over 4

=item as_string() : Str

returns "current" string.

following two variables are equivalent:

    my $a = $str->as_string();
    my $b = "$str";

=item set( $val : Str ) : String::Incremental

sets to $val.

tying with String::Incremental, assignment syntax is available as synonym of this method:

    tie my $str, 'String::Incremental', (
        format => 'foo-%2=-%=',
        orders => [ [0..2], 'abcd' ],
    );

    $str = 'foo-22-d';  # same as `$str->set( 'foo-22-d' )`
    print "$str";  # -> 'foo-22-d';

=item increment() : Str

increases positional state of order and returns its character.

following two operation are equivalent:

    $str->increment();
    $str++;

=item decrement() : Str

decreases positional state of order and returns its character.

following two operation are equivalent:

    $str->decrement();
    $str--;

=back

=head1 FUNCTIONS

=over 4

=item incremental_string( $format, @orders ) : String::Incremental

another way to construct String::Incremental instance.

this function is not exported automatically, you need to export manually:

    use String::Incremental qw( incremental_string );

=back

=head1 LICENSE

Copyright (C) issm.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

issm E<lt>issmxx@gmail.comE<gt>

=cut
